doc/man/man3/module.3
.TH "module" 3 "Rabu 8 Februari 2017" "Version 1.0.2-4" "Sarasvati" \" -*- nroff -*-
.ad l
.nh
.SH NAME
module \- The module class\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <module\&.h>\fP
.SS "Fungsi Anggota Publik"
.SH "Keterangan Lengkap"
.PP 
The module class\&. 
.SH "Dokumentasi Konstruktor & Destruktor"
.PP 
.SS "\fBmodule\fP ()"

.PP
module 
.PP
.nf
4 {
5 
6 }
.fi
.SH "Dokumentasi Anggota: Fungsi"
.PP 
.SS "void read_module (struct \fBt_module\fP * tmodule, QString addressModule)"

.PP
read_module 
.PP
\fBParameter:\fP
.RS 4
\fItmodule\fP 
.br
\fIaddressModule\fP 
.RE
.PP

.PP
.nf
144                                                                        {
145     QString pth = addressModule;
146     cryp code; code\&.decryp(pth);
147     QString temp;
148     QSettings sett(pth, QSettings::IniFormat);
149 
150     tmodule->flag_active = sett\&.value("MODULE/ACTIVE")\&.toInt();
151     strcpy(tmodule->module_name, sett\&.value("MODULE/MODULE_NAME")\&.toString()\&.toLatin1());
152     strcpy(tmodule->serial_number, sett\&.value("MODULE/SN")\&.toString()\&.toLatin1());
153     strcpy(tmodule->ip_address, sett\&.value("MODULE/IP_ADDRESS")\&.toString()\&.toLatin1());
154     strcpy(tmodule->server_address, sett\&.value("MODULE/SERVER_ADDRESS")\&.toString()\&.toLatin1());
155     strcpy(tmodule->file_address, sett\&.value("MODULE/FILE_ADDRESS")\&.toString()\&.toLatin1());
156     tmodule->flag_webclient = sett\&.value("MODULE/FLAG_WEBCLIENT")\&.toInt();
157     strcpy(tmodule->status_webclient, sett\&.value("MODULE/WEBCLIENT")\&.toString()\&.toLatin1());
158     tmodule->interval_relay = sett\&.value("MODULE/INTERVAL_RELAY")\&.toInt();
159     tmodule->interval_data = sett\&.value("MODULE/INTERVAL_DATA")\&.toInt();
160     tmodule->flag_mode_http = sett\&.value("MODULE/FLAG_MODE_HTTP")\&.toInt();
161     strcpy(tmodule->mode_http, sett\&.value("MODULE/MODE_HTTP")\&.toString()\&.toLatin1());
162     tmodule->port = sett\&.value("MODULE/PORT")\&.toInt();
163     tmodule->utc = sett\&.value("MODULE/UTC")\&.toInt();
164 
165     tmodule->flag_dual_gsm = sett\&.value("MODULE/FLAG_DUAL_GSM")\&.toInt();
166 
167     tmodule->jml_input_digital = sett\&.value("INPUT/JML_INPUT_DIGITAL")\&.toInt();
168     tmodule->jml_input_analog = sett\&.value("INPUT/JML_INPUT_ANALOG")\&.toInt();
169     tmodule->Input\&.clear(); tmodule->InputName\&.clear();
170     int j = 0;
171     for (int i = 1; i <= tmodule->jml_input_digital + tmodule->jml_input_analog; i++) {
172         if (i <= tmodule->jml_input_digital) {
173             temp\&.sprintf("INPUT/KANAL_%d", i);
174             tmodule->Input\&.insert(j,sett\&.value(temp)\&.toString());
175             temp\&.sprintf("INPUT/KANAL_NAME_%d", i);
176             tmodule->InputName\&.insert(j,sett\&.value(temp)\&.toString());
177             j++;
178         } else {
179 //            temp\&.sprintf("INPUT/KANAL_%d", i+(DATA_PERIOD-tmodule->jml_input_digital));
180             temp\&.sprintf("INPUT/KANAL_%d", i);
181             tmodule->Input\&.insert(j,sett\&.value(temp)\&.toString());
182 //            temp\&.sprintf("INPUT/KANAL_NAME_%d", i+(DATA_PERIOD-tmodule->jml_input_digital));
183             temp\&.sprintf("INPUT/KANAL_NAME_%d", i);
184             tmodule->InputName\&.insert(j,sett\&.value(temp)\&.toString());
185             j++;
186         }
187     }
188 
189     tmodule->jml_output = sett\&.value("OUTPUT/JML_OUTPUT")\&.toInt();
190     tmodule->Output\&.clear(); tmodule->OutputName\&.clear();
191     for (int i = 1; i <= tmodule->jml_output; i++) {
192         temp\&.sprintf("OUTPUT/RELAY_%d", i);
193         tmodule->Output\&.insert(i-1,sett\&.value(temp)\&.toString());
194         temp\&.sprintf("OUTPUT/RELAY_NAME_%d", i);
195         tmodule->OutputName\&.insert(i-1,sett\&.value(temp)\&.toString());
196     }
197 
198     tmodule->flag_gsm_1 = sett\&.value("GSM_1/FLAG_GSM_1")\&.toInt();
199     strcpy(tmodule->name_gsm_1, sett\&.value("GSM_1/GSM_NAME_1")\&.toString()\&.toLatin1());
200     strcpy(tmodule->device_name_gsm_1, sett\&.value("GSM_1/DEVICE_NAME_1")\&.toString()\&.toLatin1());
201     tmodule->flag_status_active_gsm_1 = sett\&.value("GSM_1/FLAG_STATUS_GSM_1")\&.toInt();
202     strcpy(tmodule->status_gsm_1, sett\&.value("GSM_1/STATUS_GSM_1")\&.toString()\&.toLatin1());
203     tmodule->flag_com_gsm_1 = sett\&.value("GSM_1/FLAG_COM_GSM_1")\&.toInt();
204     strcpy(tmodule->com_gsm_1, sett\&.value("GSM_1/COM_GSM_1")\&.toString()\&.toLatin1());
205     strcpy(tmodule->number_gsm_1, sett\&.value("GSM_1/NUMBER_GSM_1")\&.toString()\&.toLatin1());
206     strcpy(tmodule->apn_gsm_1, sett\&.value("GSM_1/APN_GSM_1")\&.toString()\&.toLatin1());
207     if (QString(tmodule->apn_gsm_1) == "") strcpy(tmodule->apn_gsm_1,"-");
208     strcpy(tmodule->user_gsm_1, sett\&.value("GSM_1/USER_GSM_1")\&.toString()\&.toLatin1());
209     if (QString(tmodule->user_gsm_1) == "") strcpy(tmodule->user_gsm_1,"-");
210     strcpy(tmodule->passwd_gsm_1, sett\&.value("GSM_1/PASSWD_GSM_1")\&.toString()\&.toLatin1());
211     if (QString(tmodule->passwd_gsm_1) == "") strcpy(tmodule->passwd_gsm_1,"-");
212 
213     tmodule->flag_gsm_2 = sett\&.value("GSM_2/FLAG_GSM_2")\&.toInt();
214     strcpy(tmodule->name_gsm_2, sett\&.value("GSM_2/GSM_NAME_2")\&.toString()\&.toLatin1());
215     strcpy(tmodule->device_name_gsm_2, sett\&.value("GSM_2/DEVICE_NAME_2")\&.toString()\&.toLatin1());
216     tmodule->flag_status_active_gsm_2 = sett\&.value("GSM_2/FLAG_STATUS_GSM_2")\&.toInt();
217     strcpy(tmodule->status_gsm_2, sett\&.value("GSM_2/STATUS_GSM_2")\&.toString()\&.toLatin1());
218     tmodule->flag_com_gsm_2 = sett\&.value("GSM_2/FLAG_COM_GSM_2")\&.toInt();
219     strcpy(tmodule->com_gsm_2, sett\&.value("GSM_2/COM_GSM_2")\&.toString()\&.toLatin1());
220     strcpy(tmodule->number_gsm_2, sett\&.value("GSM_2/NUMBER_GSM_2")\&.toString()\&.toLatin1());
221     strcpy(tmodule->apn_gsm_2, sett\&.value("GSM_2/APN_GSM_2")\&.toString()\&.toLatin1());
222     if (QString(tmodule->apn_gsm_2) == "") strcpy(tmodule->apn_gsm_2,"-");
223     strcpy(tmodule->user_gsm_2, sett\&.value("GSM_2/USER_GSM_2")\&.toString()\&.toLatin1());
224     if (QString(tmodule->user_gsm_2) == "") strcpy(tmodule->user_gsm_2,"-");
225     strcpy(tmodule->passwd_gsm_2, sett\&.value("GSM_2/PASSWD_GSM_2")\&.toString()\&.toLatin1());
226     if (QString(tmodule->passwd_gsm_2) == "") strcpy(tmodule->passwd_gsm_2,"-");
227 
228     tmodule->jml_sumber = sett\&.value("SOURCES/JML_SUMBER")\&.toInt();
229     tmodule->sumber\&.clear();
230     for (int i = 1; i <= tmodule->jml_sumber; i++) {
231         temp\&.sprintf("SOURCES/SUMBER_%d", i);
232         tmodule->sumber\&.insert(i-1,sett\&.value(temp)\&.toString());
233     }
234 
235     tmodule->jml_alarm = sett\&.value("ALARM/JML_ALARM")\&.toInt();
236     tmodule->alarm\&.clear();
237     for (int i = 1; i <= tmodule->jml_alarm; i++) {
238         temp\&.sprintf("ALARM/ALARM_%d", i);
239         tmodule->alarm\&.insert(i-1,sett\&.value(temp)\&.toString());
240     }
241 
242     tmodule->jml_data = sett\&.value("DATA/JML_DATA")\&.toInt();
243     tmodule->data\&.clear();
244     for (int i = 1; i <= tmodule->jml_data; i++) {
245         temp\&.sprintf("DATA/DATA_%d", i);
246         tmodule->data\&.insert(i-1,sett\&.value(temp)\&.toString());
247     }
248 
249     code\&.encryp(pth);
250 }
.fi
.SS "void save_as_module (struct \fBt_module\fP * tmodule, QString address)"

.PP
save_as_module 
.PP
\fBParameter:\fP
.RS 4
\fItmodule\fP 
.br
\fIaddress\fP 
.RE
.PP

.PP
.nf
252                                                                     {
253     QStringList temp1; QString temp2; int index;
254     QString pth = address;
255     QSettings sett(pth, QSettings::IniFormat);
256 
257     sett\&.beginGroup( "MODULE" );
258     sett\&.setValue("ACTIVE", tmodule->flag_active);
259     sett\&.setValue("MODULE_NAME", tmodule->module_name);
260     sett\&.setValue("SN", tmodule->serial_number);
261     sett\&.setValue("FLAG_DUAL_GSM", tmodule->flag_dual_gsm);
262     sett\&.setValue("IP_ADDRESS", tmodule->ip_address);
263     sett\&.setValue("SERVER_ADDRESS", tmodule->server_address);
264     sett\&.setValue("FILE_ADDRESS", tmodule->file_address);
265     sett\&.setValue("FLAG_WEBCLIENT", tmodule->flag_webclient);
266     sett\&.setValue("WEBCLIENT", tmodule->status_webclient);
267     sett\&.endGroup();
268 
269     sett\&.beginGroup( "INPUT" );
270     for (int i = 0; i < tmodule->Input\&.length(); i++) {
271         temp2 = tmodule->Input\&.at(i);
272         temp1 = temp2\&.split(';');
273         temp2 = "KANAL_" + temp1\&.at(1);
274         sett\&.setValue(temp2, tmodule->Input\&.at(i));
275     }
276     for (int i = 0; i < tmodule->InputName\&.length(); i++) {
277         temp2 = tmodule->Input\&.at(i);
278         temp1 = temp2\&.split(';');
279         temp2 = "KANAL_NAME_" + temp1\&.at(1);
280         sett\&.setValue(temp2, tmodule->InputName\&.at(i));
281     }
282     sett\&.setValue("JML_INPUT_DIGITAL", tmodule->jml_input_digital);
283     sett\&.setValue("JML_INPUT_ANALOG", tmodule->jml_input_analog);
284     sett\&.endGroup();
285 
286     sett\&.beginGroup( "OUTPUT" );
287     for (int i = 0; i < tmodule->Output\&.length(); i++) {
288         temp2 = tmodule->Output\&.at(i);
289         temp1 = temp2\&.split(';');
290         temp2 = "RELAY_" + temp1\&.at(1);
291         sett\&.setValue(temp2, tmodule->Output\&.at(i));
292         temp2 = "RELAY_NAME_" + temp1\&.at(1);
293         sett\&.setValue(temp2, tmodule->OutputName\&.at(i));
294     }
295     sett\&.setValue("JML_OUTPUT", tmodule->jml_output);
296     sett\&.endGroup();
297 
298     sett\&.beginGroup( "GSM_1" );
299     sett\&.setValue("FLAG_GSM_1", tmodule->flag_gsm_1);
300     sett\&.setValue("GSM_NAME_1", tmodule->name_gsm_1);
301     sett\&.setValue("DEVICE_NAME_1", tmodule->device_name_gsm_1);
302     sett\&.setValue("FLAG_STATUS_GSM_1", tmodule->flag_status_active_gsm_1);
303     sett\&.setValue("STATUS_GSM_1", tmodule->status_gsm_1);
304     sett\&.setValue("FLAG_COM_GSM_1", tmodule->flag_com_gsm_1);
305     sett\&.setValue("COM_GSM_1", tmodule->com_gsm_1);
306     sett\&.setValue("NUMBER_GSM_1", tmodule->number_gsm_1);
307     sett\&.setValue("APN_GSM_1", tmodule->apn_gsm_1);
308     sett\&.setValue("USER_GSM_1", tmodule->user_gsm_1);
309     sett\&.setValue("PASSWD_GSM_1", tmodule->passwd_gsm_1);
310     sett\&.endGroup();
311 
312     sett\&.beginGroup( "GSM_2" );
313     if (tmodule->flag_dual_gsm == 1) {
314         sett\&.setValue("FLAG_GSM_2", tmodule->flag_gsm_2);
315         sett\&.setValue("GSM_NAME_2", tmodule->name_gsm_2);
316         sett\&.setValue("DEVICE_NAME_2", tmodule->device_name_gsm_2);
317         sett\&.setValue("FLAG_STATUS_GSM_2", tmodule->flag_status_active_gsm_2);
318         sett\&.setValue("STATUS_GSM_2", tmodule->status_gsm_2);
319         sett\&.setValue("FLAG_COM_GSM_2", tmodule->flag_com_gsm_2);
320         sett\&.setValue("COM_GSM_2", tmodule->com_gsm_2);
321         sett\&.setValue("NUMBER_GSM_2", tmodule->number_gsm_2);

model/module.cpp
#include "module.h"

module::module()
{

}

void module::write_module(struct t_module *tmodule){

    QStringList temp1; QString temp2; int index;
    QString pth;
    pth.sprintf(".RTUdata/module/m_%s.dbe",tmodule->module_name);
    QSettings sett(pth, QSettings::IniFormat);

    sett.beginGroup( "MODULE" );
    sett.setValue("ACTIVE", tmodule->flag_active);
    sett.setValue("MODULE_NAME", tmodule->module_name);
    sett.setValue("SN", tmodule->serial_number);
    sett.setValue("FLAG_DUAL_GSM", tmodule->flag_dual_gsm);
    sett.setValue("IP_ADDRESS", tmodule->ip_address);
    sett.setValue("SERVER_ADDRESS", tmodule->server_address);
    sett.setValue("FILE_ADDRESS", tmodule->file_address);
    sett.setValue("FLAG_WEBCLIENT", tmodule->flag_webclient);
    if (tmodule->flag_webclient == 0) {
        strcpy(tmodule->status_webclient, "NOT ACTIVE");
        sett.setValue("WEBCLIENT", tmodule->status_webclient);
    } else {
        strcpy(tmodule->status_webclient, "ACTIVE");
        sett.setValue("WEBCLIENT", tmodule->status_webclient);
    }
    sett.setValue("INTERVAL_RELAY", tmodule->interval_relay);
    sett.setValue("INTERVAL_DATA", tmodule->interval_data);
    sett.setValue("FLAG_MODE_HTTP", tmodule->mode_http);
    if (tmodule->flag_mode_http == 0) {
        strcpy(tmodule->mode_http, "GET");
        sett.setValue("MODE_HTTP", tmodule->mode_http);
    } else if (tmodule->flag_mode_http == 1) {
        strcpy(tmodule->mode_http, "POST");
        sett.setValue("MODE_HTTP", tmodule->mode_http);
    }
    sett.setValue("PORT", tmodule->port);
    sett.setValue("UTC", tmodule->utc);
    sett.endGroup();

    sett.beginGroup( "INPUT" );
    for (int i = 0; i < tmodule->Input.length(); i++) {
        temp2 = tmodule->Input.at(i);
        temp1 = temp2.split(';');
        temp2 = "KANAL_" + temp1.at(1);
        sett.setValue(temp2, tmodule->Input.at(i));
    }
    for (int i = 0; i < tmodule->InputName.length(); i++) {
        temp2 = tmodule->Input.at(i);
        temp1 = temp2.split(';');
        temp2 = "KANAL_NAME_" + temp1.at(1);
        sett.setValue(temp2, tmodule->InputName.at(i));
    }
    sett.setValue("JML_INPUT_DIGITAL", tmodule->jml_input_digital);
    sett.setValue("JML_INPUT_ANALOG", tmodule->jml_input_analog);
    sett.endGroup();

    sett.beginGroup( "OUTPUT" );
    for (int i = 0; i < tmodule->Output.length(); i++) {
        temp2 = tmodule->Output.at(i);
        temp1 = temp2.split(';');
        temp2 = "RELAY_" + temp1.at(1);
        sett.setValue(temp2, tmodule->Output.at(i));
        temp2 = "RELAY_NAME_" + temp1.at(1);
        sett.setValue(temp2, tmodule->OutputName.at(i));
    }
    sett.setValue("JML_OUTPUT", tmodule->jml_output);
    sett.endGroup();

    sett.beginGroup( "GSM_1" );
    sett.setValue("FLAG_GSM_1", tmodule->flag_gsm_1);
    sett.setValue("GSM_NAME_1", tmodule->name_gsm_1);
    sett.setValue("DEVICE_NAME_1", tmodule->device_name_gsm_1);
    sett.setValue("FLAG_STATUS_GSM_1", tmodule->flag_status_active_gsm_1);
    sett.setValue("STATUS_GSM_1", tmodule->status_gsm_1);
    sett.setValue("FLAG_COM_GSM_1", tmodule->flag_com_gsm_1);
    sett.setValue("COM_GSM_1", tmodule->com_gsm_1);
    sett.setValue("NUMBER_GSM_1", tmodule->number_gsm_1);
    sett.setValue("APN_GSM_1", tmodule->apn_gsm_1);
    sett.setValue("USER_GSM_1", tmodule->user_gsm_1);
    sett.setValue("PASSWD_GSM_1", tmodule->passwd_gsm_1);
    sett.endGroup();

    sett.beginGroup( "GSM_2" );
    if (tmodule->flag_dual_gsm == 1) {
        sett.setValue("FLAG_GSM_2", tmodule->flag_gsm_2);
        sett.setValue("GSM_NAME_2", tmodule->name_gsm_2);
        sett.setValue("DEVICE_NAME_2", tmodule->device_name_gsm_2);
        sett.setValue("FLAG_STATUS_GSM_2", tmodule->flag_status_active_gsm_2);
        sett.setValue("STATUS_GSM_2", tmodule->status_gsm_2);
        sett.setValue("FLAG_COM_GSM_2", tmodule->flag_com_gsm_2);
        sett.setValue("COM_GSM_2", tmodule->com_gsm_2);
        sett.setValue("NUMBER_GSM_2", tmodule->number_gsm_2);
        sett.setValue("APN_GSM_2", tmodule->apn_gsm_2);
        sett.setValue("USER_GSM_2", tmodule->user_gsm_2);
        sett.setValue("PASSWD_GSM_2", tmodule->passwd_gsm_2);
    } else {
        sett.setValue("FLAG_GSM_2", "");
        sett.setValue("GSM_NAME_2", "");
        sett.setValue("DEVICE_NAME_2", "");
        sett.setValue("FLAG_STATUS_GSM_2", "");
        sett.setValue("STATUS_GSM_2", "");
        sett.setValue("FLAG_COM_GSM_2", "");
        sett.setValue("COM_GSM_2", "");
        sett.setValue("NUMBER_GSM_2", "");
        sett.setValue("APN_GSM_2", "");
        sett.setValue("USER_GSM_2", "");
        sett.setValue("PASSWD_GSM_2", "");
    }
    sett.endGroup();

    sett.beginGroup( "SOURCES" );
    index = tmodule->sumber.length();
    for (int i = 0; i < index; i++) {
        temp2.sprintf("SUMBER_%d", i+1);
        sett.setValue(temp2, tmodule->sumber.at(i));
    }
    sett.setValue("JML_SUMBER", tmodule->jml_sumber);
    sett.endGroup();

    sett.beginGroup( "ALARM" );
    index = tmodule->alarm.length();
    for (int i = 0; i < index; i++) {
        temp2.sprintf("ALARM_%d", i+1);
        sett.setValue(temp2, tmodule->alarm.at(i));
    }
    sett.setValue("JML_ALARM", tmodule->jml_alarm);
    sett.endGroup();

    sett.beginGroup( "DATA" );
    index = tmodule->data.length();
    for (int i = 0; i < index; i++) {
        temp2.sprintf("DATA_%d", i+1);
        sett.setValue(temp2, tmodule->data.at(i));
    }
    sett.setValue("JML_DATA", tmodule->jml_data);
    sett.endGroup();
}

void module::read_module(struct t_module *tmodule, QString addressModule){
    QString pth = addressModule;
    cryp code; code.decryp(pth);
    QString temp;
    QSettings sett(pth, QSettings::IniFormat);

    tmodule->flag_active = sett.value("MODULE/ACTIVE").toInt();
    strcpy(tmodule->module_name, sett.value("MODULE/MODULE_NAME").toString().toLatin1());
    strcpy(tmodule->serial_number, sett.value("MODULE/SN").toString().toLatin1());
    strcpy(tmodule->ip_address, sett.value("MODULE/IP_ADDRESS").toString().toLatin1());
    strcpy(tmodule->server_address, sett.value("MODULE/SERVER_ADDRESS").toString().toLatin1());
    strcpy(tmodule->file_address, sett.value("MODULE/FILE_ADDRESS").toString().toLatin1());
    tmodule->flag_webclient = sett.value("MODULE/FLAG_WEBCLIENT").toInt();
    strcpy(tmodule->status_webclient, sett.value("MODULE/WEBCLIENT").toString().toLatin1());
    tmodule->interval_relay = sett.value("MODULE/INTERVAL_RELAY").toInt();
    tmodule->interval_data = sett.value("MODULE/INTERVAL_DATA").toInt();
    tmodule->flag_mode_http = sett.value("MODULE/FLAG_MODE_HTTP").toInt();
    strcpy(tmodule->mode_http, sett.value("MODULE/MODE_HTTP").toString().toLatin1());
    tmodule->port = sett.value("MODULE/PORT").toInt();
    tmodule->utc = sett.value("MODULE/UTC").toInt();

    tmodule->flag_dual_gsm = sett.value("MODULE/FLAG_DUAL_GSM").toInt();

    tmodule->jml_input_digital = sett.value("INPUT/JML_INPUT_DIGITAL").toInt();
    tmodule->jml_input_analog = sett.value("INPUT/JML_INPUT_ANALOG").toInt();
    tmodule->Input.clear(); tmodule->InputName.clear();
    int j = 0;
    for (int i = 1; i <= tmodule->jml_input_digital + tmodule->jml_input_analog; i++) {
        if (i <= tmodule->jml_input_digital) {
            temp.sprintf("INPUT/KANAL_%d", i);
            tmodule->Input.insert(j,sett.value(temp).toString());
            temp.sprintf("INPUT/KANAL_NAME_%d", i);
            tmodule->InputName.insert(j,sett.value(temp).toString());
            j++;
        } else {
//            temp.sprintf("INPUT/KANAL_%d", i+(DATA_PERIOD-tmodule->jml_input_digital));
            temp.sprintf("INPUT/KANAL_%d", i);
            tmodule->Input.insert(j,sett.value(temp).toString());
//            temp.sprintf("INPUT/KANAL_NAME_%d", i+(DATA_PERIOD-tmodule->jml_input_digital));
            temp.sprintf("INPUT/KANAL_NAME_%d", i);
            tmodule->InputName.insert(j,sett.value(temp).toString());
            j++;
        }
    }

    tmodule->jml_output = sett.value("OUTPUT/JML_OUTPUT").toInt();
    tmodule->Output.clear(); tmodule->OutputName.clear();
    for (int i = 1; i <= tmodule->jml_output; i++) {
        temp.sprintf("OUTPUT/RELAY_%d", i);
        tmodule->Output.insert(i-1,sett.value(temp).toString());
        temp.sprintf("OUTPUT/RELAY_NAME_%d", i);
        tmodule->OutputName.insert(i-1,sett.value(temp).toString());
    }

    tmodule->flag_gsm_1 = sett.value("GSM_1/FLAG_GSM_1").toInt();
    strcpy(tmodule->name_gsm_1, sett.value("GSM_1/GSM_NAME_1").toString().toLatin1());
    strcpy(tmodule->device_name_gsm_1, sett.value("GSM_1/DEVICE_NAME_1").toString().toLatin1());
    tmodule->flag_status_active_gsm_1 = sett.value("GSM_1/FLAG_STATUS_GSM_1").toInt();
    strcpy(tmodule->status_gsm_1, sett.value("GSM_1/STATUS_GSM_1").toString().toLatin1());
    tmodule->flag_com_gsm_1 = sett.value("GSM_1/FLAG_COM_GSM_1").toInt();
    strcpy(tmodule->com_gsm_1, sett.value("GSM_1/COM_GSM_1").toString().toLatin1());
    strcpy(tmodule->number_gsm_1, sett.value("GSM_1/NUMBER_GSM_1").toString().toLatin1());
    strcpy(tmodule->apn_gsm_1, sett.value("GSM_1/APN_GSM_1").toString().toLatin1());
    if (QString(tmodule->apn_gsm_1) == "") strcpy(tmodule->apn_gsm_1,"-");
    strcpy(tmodule->user_gsm_1, sett.value("GSM_1/USER_GSM_1").toString().toLatin1());
    if (QString(tmodule->user_gsm_1) == "") strcpy(tmodule->user_gsm_1,"-");
    strcpy(tmodule->passwd_gsm_1, sett.value("GSM_1/PASSWD_GSM_1").toString().toLatin1());
    if (QString(tmodule->passwd_gsm_1) == "") strcpy(tmodule->passwd_gsm_1,"-");

    tmodule->flag_gsm_2 = sett.value("GSM_2/FLAG_GSM_2").toInt();
    strcpy(tmodule->name_gsm_2, sett.value("GSM_2/GSM_NAME_2").toString().toLatin1());
    strcpy(tmodule->device_name_gsm_2, sett.value("GSM_2/DEVICE_NAME_2").toString().toLatin1());
    tmodule->flag_status_active_gsm_2 = sett.value("GSM_2/FLAG_STATUS_GSM_2").toInt();
    strcpy(tmodule->status_gsm_2, sett.value("GSM_2/STATUS_GSM_2").toString().toLatin1());
    tmodule->flag_com_gsm_2 = sett.value("GSM_2/FLAG_COM_GSM_2").toInt();
    strcpy(tmodule->com_gsm_2, sett.value("GSM_2/COM_GSM_2").toString().toLatin1());
    strcpy(tmodule->number_gsm_2, sett.value("GSM_2/NUMBER_GSM_2").toString().toLatin1());
    strcpy(tmodule->apn_gsm_2, sett.value("GSM_2/APN_GSM_2").toString().toLatin1());
    if (QString(tmodule->apn_gsm_2) == "") strcpy(tmodule->apn_gsm_2,"-");
    strcpy(tmodule->user_gsm_2, sett.value("GSM_2/USER_GSM_2").toString().toLatin1());
    if (QString(tmodule->user_gsm_2) == "") strcpy(tmodule->user_gsm_2,"-");
    strcpy(tmodule->passwd_gsm_2, sett.value("GSM_2/PASSWD_GSM_2").toString().toLatin1());
    if (QString(tmodule->passwd_gsm_2) == "") strcpy(tmodule->passwd_gsm_2,"-");

    tmodule->jml_sumber = sett.value("SOURCES/JML_SUMBER").toInt();
    tmodule->sumber.clear();
    for (int i = 1; i <= tmodule->jml_sumber; i++) {
        temp.sprintf("SOURCES/SUMBER_%d", i);
        tmodule->sumber.insert(i-1,sett.value(temp).toString());
    }

    tmodule->jml_alarm = sett.value("ALARM/JML_ALARM").toInt();
    tmodule->alarm.clear();
    for (int i = 1; i <= tmodule->jml_alarm; i++) {
        temp.sprintf("ALARM/ALARM_%d", i);
        tmodule->alarm.insert(i-1,sett.value(temp).toString());
    }

    tmodule->jml_data = sett.value("DATA/JML_DATA").toInt();
    tmodule->data.clear();
    for (int i = 1; i <= tmodule->jml_data; i++) {
        temp.sprintf("DATA/DATA_%d", i);
        tmodule->data.insert(i-1,sett.value(temp).toString());
    }

    code.encryp(pth);
}

void module::save_as_module(struct t_module *tmodule, QString address){
    QStringList temp1; QString temp2; int index;
    QString pth = address;
    QSettings sett(pth, QSettings::IniFormat);

    sett.beginGroup( "MODULE" );
    sett.setValue("ACTIVE", tmodule->flag_active);
    sett.setValue("MODULE_NAME", tmodule->module_name);
    sett.setValue("SN", tmodule->serial_number);
    sett.setValue("FLAG_DUAL_GSM", tmodule->flag_dual_gsm);
    sett.setValue("IP_ADDRESS", tmodule->ip_address);
    sett.setValue("SERVER_ADDRESS", tmodule->server_address);
    sett.setValue("FILE_ADDRESS", tmodule->file_address);
    sett.setValue("FLAG_WEBCLIENT", tmodule->flag_webclient);
    sett.setValue("WEBCLIENT", tmodule->status_webclient);
    sett.endGroup();

    sett.beginGroup( "INPUT" );
    for (int i = 0; i < tmodule->Input.length(); i++) {
        temp2 = tmodule->Input.at(i);
        temp1 = temp2.split(';');
        temp2 = "KANAL_" + temp1.at(1);
        sett.setValue(temp2, tmodule->Input.at(i));
    }
    for (int i = 0; i < tmodule->InputName.length(); i++) {
        temp2 = tmodule->Input.at(i);
        temp1 = temp2.split(';');
        temp2 = "KANAL_NAME_" + temp1.at(1);
        sett.setValue(temp2, tmodule->InputName.at(i));
    }
    sett.setValue("JML_INPUT_DIGITAL", tmodule->jml_input_digital);
    sett.setValue("JML_INPUT_ANALOG", tmodule->jml_input_analog);
    sett.endGroup();

    sett.beginGroup( "OUTPUT" );
    for (int i = 0; i < tmodule->Output.length(); i++) {
        temp2 = tmodule->Output.at(i);
        temp1 = temp2.split(';');
        temp2 = "RELAY_" + temp1.at(1);
        sett.setValue(temp2, tmodule->Output.at(i));
        temp2 = "RELAY_NAME_" + temp1.at(1);
        sett.setValue(temp2, tmodule->OutputName.at(i));
    }
    sett.setValue("JML_OUTPUT", tmodule->jml_output);
    sett.endGroup();

    sett.beginGroup( "GSM_1" );
    sett.setValue("FLAG_GSM_1", tmodule->flag_gsm_1);
    sett.setValue("GSM_NAME_1", tmodule->name_gsm_1);
    sett.setValue("DEVICE_NAME_1", tmodule->device_name_gsm_1);
    sett.setValue("FLAG_STATUS_GSM_1", tmodule->flag_status_active_gsm_1);
    sett.setValue("STATUS_GSM_1", tmodule->status_gsm_1);
    sett.setValue("FLAG_COM_GSM_1", tmodule->flag_com_gsm_1);
    sett.setValue("COM_GSM_1", tmodule->com_gsm_1);
    sett.setValue("NUMBER_GSM_1", tmodule->number_gsm_1);
    sett.setValue("APN_GSM_1", tmodule->apn_gsm_1);
    sett.setValue("USER_GSM_1", tmodule->user_gsm_1);
    sett.setValue("PASSWD_GSM_1", tmodule->passwd_gsm_1);
    sett.endGroup();

    sett.beginGroup( "GSM_2" );
    if (tmodule->flag_dual_gsm == 1) {
        sett.setValue("FLAG_GSM_2", tmodule->flag_gsm_2);
        sett.setValue("GSM_NAME_2", tmodule->name_gsm_2);
        sett.setValue("DEVICE_NAME_2", tmodule->device_name_gsm_2);
        sett.setValue("FLAG_STATUS_GSM_2", tmodule->flag_status_active_gsm_2);
        sett.setValue("STATUS_GSM_2", tmodule->status_gsm_2);
        sett.setValue("FLAG_COM_GSM_2", tmodule->flag_com_gsm_2);
        sett.setValue("COM_GSM_2", tmodule->com_gsm_2);
        sett.setValue("NUMBER_GSM_2", tmodule->number_gsm_2);
        sett.setValue("APN_GSM_2", tmodule->apn_gsm_2);
        sett.setValue("USER_GSM_2", tmodule->user_gsm_2);
        sett.setValue("PASSWD_GSM_2", tmodule->passwd_gsm_2);
    } else {
        sett.setValue("FLAG_GSM_2", "");
        sett.setValue("GSM_NAME_2", "");
        sett.setValue("DEVICE_NAME_2", "");
        sett.setValue("FLAG_STATUS_GSM_2", "");
        sett.setValue("STATUS_GSM_2", "");
        sett.setValue("FLAG_COM_GSM_2", "");
        sett.setValue("COM_GSM_2", "");
        sett.setValue("NUMBER_GSM_2", "");
        sett.setValue("APN_GSM_2", "");
        sett.setValue("USER_GSM_2", "");
        sett.setValue("PASSWD_GSM_2", "");
    }
    sett.endGroup();

    sett.beginGroup( "SOURCES" );
    index = tmodule->sumber.length();
    for (int i = 0; i < index; i++) {
        temp2.sprintf("SUMBER_%d", i+1);
        sett.setValue(temp2, tmodule->sumber.at(i));
    }
    sett.setValue("JML_SUMBER", tmodule->jml_sumber);
    sett.endGroup();

    sett.beginGroup( "ALARM" );
    index = tmodule->alarm.length();
    for (int i = 0; i < index; i++) {
        temp2.sprintf("ALARM_%d", i+1);
        sett.setValue(temp2, tmodule->alarm.at(i));
    }
    sett.setValue("JML_ALARM", tmodule->jml_alarm);
    sett.endGroup();

    sett.beginGroup( "DATA" );
    index = tmodule->data.length();
    for (int i = 0; i < index; i++) {
        temp2.sprintf("DATA_%d", i+1);
        sett.setValue(temp2, tmodule->data.at(i));
    }
    sett.setValue("JML_DATA", tmodule->jml_data);
    sett.endGroup();
}
